### ISSUE TITLE
Papyrus - Stack too deep (infinite recursion likely)

### ISSUE THREAD

**Issue by Scrivener07** [Scrivener07/FO4_Games#48]
After playing blackjack for some time I get two errors in the papyrus log. The papyrus errors are `Stack too deep` and `Assigning None to a non-object variable`. When these errors occur the buttonhint script will still respond to registered key press events but the buttonhint UI HUD element will no longer be visible.

[Papyrus Error - Stack Too Deep](https://www.creationkit.com/fallout4/index.php?title=Papyrus_Runtime_Errors#.22Stack_too_deep_.28infinite_recursion_likely.29_-_aborting_call_and_returning_None.22)
**"Stack too deep (infinite recursion likely) - aborting call and returning None"**
Too many functions have been called without first returning from them. This usually happens due to infinite recursion in a script, but could also appear in highly complicated scripts that have lots of deep function calls. Try simplifying the script, removing the recursion, and making sure that any recursion has a termination condition.

```
[01/12/2018 - 02:03:45AM] error: Stack too deep (infinite recursion likely) - aborting call and returning None
stack:
	[Games_Script_Blackjack_PlayerHuman (08002761)].games:shared:ui:buttonhint.GetMember() - "D:\Games\Steam\SteamApps\common\Fallout 4\Data\Scripts\Source\FO4_Games\Games\Shared\UI\Display.psc" Line 51
	<unknown self>.games:shared:ui:framework.DisplaySetVisible() - "D:\Games\Steam\SteamApps\common\Fallout 4\Data\Scripts\Source\FO4_Games\Games\Shared\UI\Framework.psc" Line 165
	[Games_Script_Blackjack_PlayerHuman (08002761)].games:shared:ui:buttonhint.Visible() - "D:\Games\Steam\SteamApps\common\Fallout 4\Data\Scripts\Source\FO4_Games\Games\Shared\UI\Display.psc" Line 113
	[Games_Script_Blackjack_PlayerHuman (08002761)].games:shared:ui:buttonhint.Shown.OnBeginState() - "D:\Games\Steam\SteamApps\common\Fallout 4\Data\Scripts\Source\FO4_Games\Games\Shared\UI\ButtonHint.psc" Line 181
	
	<truncated stack>

[01/12/2018 - 02:03:45AM] warning: Assigning None to a non-object variable named "::temp9"
stack:
	[Games_Script_Blackjack_PlayerHuman (08002761)].games:shared:ui:buttonhint.GetMember() - "D:\Games\Steam\SteamApps\common\Fallout 4\Data\Scripts\Source\FO4_Games\Games\Shared\UI\Display.psc" Line 51
	<unknown self>.games:shared:ui:framework.DisplaySetVisible() - "D:\Games\Steam\SteamApps\common\Fallout 4\Data\Scripts\Source\FO4_Games\Games\Shared\UI\Framework.psc" Line 165
	[Games_Script_Blackjack_PlayerHuman (08002761)].games:shared:ui:buttonhint.Visible() - "D:\Games\Steam\SteamApps\common\Fallout 4\Data\Scripts\Source\FO4_Games\Games\Shared\UI\Display.psc" Line 113
	[Games_Script_Blackjack_PlayerHuman (08002761)].games:shared:ui:buttonhint.Shown.OnBeginState() - "D:\Games\Steam\SteamApps\common\Fallout 4\Data\Scripts\Source\FO4_Games\Games\Shared\UI\ButtonHint.psc" Line 181
	
	<truncated stack>
```

This problem applies to at least commit at https://github.com/Scrivener07/FO4_Games/tree/8784776a121c8c23106cef57153b2a570d55ee51

**Comment by Scrivener07** [comment]
A common factor Ive noticed is that the error happens when the player has run out of caps to play the game with.

**Comment by Scrivener07** [comment]
The default maximum stack size seems to be 100 according to the [CK wiki](https://www.creationkit.com/fallout4/index.php?title=INI_Settings_(Papyrus)#Defaults).

**Comment by Scrivener07** [comment]
I used the games console to [toggle papyrus global profiling](https://www.creationkit.com/fallout4/index.php?title=TogglePapyrusGlobalProfiler) with `TPGP`.

In the `Global.0.log` file I can see where one of my script stack depth exceeds 99.

**Comment by TheWhiteCollarPlayers** [comment]
Notes from our discussion last night on things to try to work this out:
-"flattening" of code
-limiting number of possible threads that can run through the code
-checking for functions called inside their own definitions

**Comment by Scrivener07** [comment]
Good list. I also need to narrow down the profiler to just this projects scripts. Im pretty new to using a profiler like this so its been a tinker.

**Comment by Scrivener07** [comment]
This is potenially fixed with [Games:Blackjack:Main.RequestState(ScriptObject, int)](https://github.com/Scrivener07/FO4_Games/blob/235f1977d6245ebe4d611a69f517f3dbb02f45ad/Data/Scripts/Source/FO4_Games/Games/Blackjack/Main.psc#L286)


@TheWhiteCollarPlayers your instincts were correct the whole time. `GotoState` and `OnStateBegin` didnt work how I originally thought with the whole starting a new thread for the event.
The `OnActivate` WAS waiting for the game to return from all the functions. With every match the call stack was getting deeper and deeper starting from `OnActivate` until the crash happened.
I fixed it by calling state changes from within an `OnTimer` event which DOES start with its own thread.

Like this.
```
Function DerpDerp()
    ; lets request the state instead of changing directly
    RequestState(self, WageringID)
EndFunction

int IdlingID   = 10 const
int StartingID = 20 const
int WageringID = 30 const
int DealingID  = 40 const
int PlayingID  = 50 const
int ScoringID  = 60 const
int ExitingID  = 70 const

bool Function RequestState(ScriptObject script, int timerID)
    script.StartTimer(0.1, timerID)
    return true
EndFunction

Event OnTimer(int aiTimerID)
    If (aiTimerID == IdlingID)
        ChangeState(self, IdlingState)
    ElseIf (aiTimerID == StartingID)
        ChangeState(self, StartingState)
    ElseIf (aiTimerID == WageringID)
        ChangeState(self, WageringState)
    ElseIf (aiTimerID == DealingID)
        ChangeState(self, DealingState)
    ElseIf (aiTimerID == PlayingID)
        ChangeState(self, PlayingState)
    ElseIf (aiTimerID == ScoringID)
        ChangeState(self, ScoringState)
    ElseIf (aiTimerID == ExitingID)
        ChangeState(self, ExitingState)
    Else
        WriteUnexpectedValue(self, "OnTimer", "aiTimerID", "The timer ID "+aiTimerID+" was unhandled.")
    EndIf
EndEvent
```

**Comment by Scrivener07** [comment]
Fixed with pull request #78
